Expose GLU activations as arguments

jaketae opened the pull request:
This PR exposes improved GLU activation functions (#47) as arguments to the main script. The added flag is 

```shell
... --glu-activation geglu
```

jaketae commented:
@thomasw21 Some additional notes:

1. I saw that you made use of enums for the rotary embedding argument, but I decided to go with raw strings + `getattr()` combo. I hope this is fine. 
2. I saw a comment that said `jit` does not play well with bp16. By default, all the GLU activation functions are `jit`'ed. If this becomes a problem, I'll add more code for non-`jit`'ed functions. 

Thanks for the review!

thomasw21 commented on megatron/model/transformer.py:
```diff
@@ -30,6 +30,7 @@
 
 import deepspeed
 
+import activations as glu_activations
```
Let's call is glu_activations then.

thomasw21 commented on megatron/model/transformer.py:
```diff
@@ -76,7 +77,9 @@ def __init__(self, init_method, output_layer_init_method):
         self.activation_func = F.gelu
-        if args.openai_gelu:
+        if args.glu_activation:
+            self.activation_func = getattr(glu_activations, args.glu_activation)
```
I personally don't like `getattr`, as I consider it bad practice. But it seems there are quite a lot of occurence of it in the codebase so I'll let you decide.

What you could have written without if, is store all activation functions in a dict. Its key is the is the string, and value is the function. I use enum as I feel string is too loose but as you want.

jaketae replied:
I also agree `getattr()` is suboptimal. I'll go with the lookup table approach, which seems like a reasonable middle ground.

thomasw21 reviewed (commented):
Thanks! Small nit, let me know what you think:
 - also could you add a test? Maybe one that just run a model with a glu config so we know it doesn't throw.
 
 About your comments:
  - I dislike `getattr` as I generally consider it bad practice, but really depends on the crowd I guess. I've seen some occurence of it in the codebase so why not?
  - Concerning that ... I don't know maybe you can test it out?

thomasw21 commented on megatron/arguments.py:
```diff
@@ -313,6 +313,11 @@ def _add_network_size_args(parser):
                        )
+    group.add_argument('--glu-activation', type=str,
+                       choices=["liglu", "geglu", "reglu", "swiglu"],
+                       default="",
```
why do you need a default?

jaketae replied:
For some reason I assumed that that having specifying a default would make it a required flag, but just realized that this is not the case, so I got rid of it. Thanks for pointing it out!

thomasw21 commented on megatron/model/transformer.py:
```diff
@@ -78,7 +78,13 @@ def __init__(self, init_method, output_layer_init_method):
         self.activation_func = F.gelu
         if args.glu_activation:
-            self.activation_func = getattr(glu_activations, args.glu_activation)
+            glu_lookup = {
```
You could make this available in glu_activation. Something like:

```python
GLU_ACTIVATIONS = { ... }
```

And basically everwhere you'd be able to call that.

jaketae replied:
That's a better idea. Just pushed a commit that implements this.

jaketae commented:
@thomasw21 Are we running experiments in `torch.bfloat16`? On my local, even using the built-in `torch.nn.functiona.gelu()` with a BF16 tensor raises this error:

```shell
(venv) jaketae:Megatron-DeepSpeed (activation-args) $ python
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> x = torch.randn(8, 100, 768)
>>> x = x.to(torch.bfloat16)
>>> x.dtype
torch.bfloat16
>>> torch.nn.functional.gelu(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jaketae/Documents/Dev/GitHub/Megatron-DeepSpeed/venv/lib/python3.7/site-packages/torch/nn/functional.py", line 1555, in gelu
    return torch._C._nn.gelu(input)
RuntimeError: "GeluKernelImpl" not implemented for 'BFloat16'
```

thomasw21 commented:
So I'm not too familiar with bf16. From what I understand it's only available for A100 (cpu support seems to be very basic). Concerning our current experiments, we're running fp16 on all experiments right now. 

JZ has very few A100 and are more experimental, so there's probably no chance we're running bf16 there. However it doesn't mean we never will. Let's make a best effort to support it.

jaketae commented:
@thomasw21 I brushed up on the things that were mentioned (e.g. `default`, `GLU_ACTIVATIONS`). As for BF16 and JIT, I think I was getting ahead of myself. Since we're using FP16 on V100s, I don't think JIT will be an issue. If it does, we can easily fix it by removing JIT.

I agree that an exhaustive high-level test that goes through the entire pipeline would be helpful. I also think @stas00 might have some ideas in mind. Maybe we can open a separate issue and continue the discussion there.

stas00 reviewed (changes requested):
Looks good!

Probably add `--glu-activation=geglu` to the main training test - at the moment it just tests that it can just run - i.e. no need for a new training test, as you're already testing the functionality elsewhere.

Like this:
https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/15/commits/8088c0fed7c350dfd50be014791cba00a6700bb7

stas00 commented:
indeed, we don't need to worry about bf16 for JZ, but we could use it on A100 if we get access.

To check for bf16 support, please see:
https://github.com/huggingface/transformers/pull/13207/commits/26eb566e43148c80d0ea098c76c3d128c0281c16

jaketae commented:
@stas00 Thanks for the review. 

1. I added `is_torch_bf16_available()` from the transformers repo. This function is called to unit test GLU activations in an environment where BF16 is available. So I assume if you run `python -m unittest tests/test_activations.py` on your local 3090, it will properly test if things work as expected.
2.  I added the activation function flag in `test_training_all()` as suggested.

Let me know if there's anything that needs to be fixed!

stas00 commented on megatron/testing_utils.py:
```diff
@@ -212,6 +213,23 @@ def torch_assert_equal(actual, expected):
+            return False
+        if int(torch.version.cuda.split(".")[0]) < 11:
+            return False
+        if not version.parse(torch.__version__) > version.parse("1.09"):
```
```suggestion
        if not version.parse(torch.__version__) >= version.parse("1.09"):
```

pt-1.9 already provides bf16 support, pt-1.10 for amp/bf16

will fix upstream.

stas00 commented on tests/test_activations.py:
```diff
@@ -40,3 +40,10 @@ def test_reglu(self):
         expected = self.x1 * F.silu(self.x2)
         torch_assert_equal(swiglu(self.x), expected)
+
+    def test_bf16_jit(self):
```
there is a better way, which properly reports skipping and not false PASSED with the current proposal when bf16 is not available.

To implement add to testing_utils.py:

```

def require_torch_bf16(test_case):
    """Decorator marking a test that requires CUDA hardware supporting bf16 and PyTorch >= 1.9."""
    if not is_torch_bf16_available():
        return unittest.skip("test requires CUDA hardware supporting bf16 and PyTorch >= 1.9")(test_case)
    else:
        return test_case
```

and then use the decorator:
```
@require_torch_bf16
def test_bf16_jit(self):
```

jaketae replied:
That's much cleaner. Thanks for the suggestion!

stas00 commented:
Here you go:

```
pyt tests/test_activations.py 
====================================================================== test session starts ======================================================================
platform linux -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /mnt/nvme1/code/huggingface, configfile: pytest.ini
plugins: dash-1.20.0, forked-1.3.0, xdist-2.3.0, instafail-0.4.2
collected 6 items                                                                                                                                               

tests/test_activations.py F

_________________________________________________________________ TestActivations.test_bf16_jit _________________________________________________________________

self = <test_activations.TestActivations testMethod=test_bf16_jit>

    @require_torch_bf16
    def test_bf16_jit(self):
        x_bf16 = self.x.to(torch.bfloat16)
        for activation_fn in GLU_ACTIVATIONS.values():
>           output = activation_fn(x_bf16)

tests/test_activations.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = RecursiveScriptModule(original_name=GEGLU)
input = (tensor([[[ 1.9297,  1.4844,  0.9023,  ...,  1.6797,  1.2812,  1.2969],
         [ 0.6094,  1.3359, -0.2314,  ..., -0....0469,  0.2852],
         [ 0.1680, -0.3379, -1.1484,  ...,  1.2500, -0.6367, -0.5156]]],
       dtype=torch.bfloat16),)
kwargs = {}, forward_call = <torch._C.ScriptMethod object at 0x7f9c9293c450>

    def _call_impl(self, *input, **kwargs):
        forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)
        # If we don't have any hooks, we want to skip the rest of the logic in
        # this function, and just call forward.
        if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
                or _global_forward_hooks or _global_forward_pre_hooks):
>           return forward_call(*input, **kwargs)
E           RuntimeError: The following operation failed in the TorchScript interpreter.
E           Traceback of TorchScript (most recent call last):
E             File "/mnt/nvme1/code/huggingface/Megatron-DeepSpeed-master/megatron/model/glu_activations.py", line 14, in forward
E                   # dim=-1 breaks in jit for pt<1.10
E                   x1, x2 = x.chunk(2, dim=(x.ndim - 1))
E                   return x1 * self.activation_fn(x2)
E                               ~~~~~~~~~~~~~~~~~~ <--- HERE
E             File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/nn/functional.py", line 1555, in gelu
E               if has_torch_function_unary(input):
E                   return handle_torch_function(gelu, (input,), input)
E               return torch._C._nn.gelu(input)
E                      ~~~~~~~~~~~~~~~~~ <--- HERE
E           RuntimeError: "GeluKernelImpl" not implemented for 'BFloat16'

/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/nn/modules/module.py:1051: RuntimeError
```

jaketae commented:
> RuntimeError: "GeluKernelImpl" not implemented for 'BFloat16'

I see you got the same error as the one I got on my local. The trace seems to suggest that we can't use BF16 with JIT due to some missing kernel implementations. Here are my thoughts:

1. We could ignore this for now since we're on V100s anyway?
2. I'll push some non-JIT'ed version of the activation functions to `glu_activations.py` so that we can use them if need be, once we have access to A100 nodes that can utilize BF16. 

What do you think?

stas00 commented:
This is unrelated to JIT, it's the same error if you remove jit
```
E       RuntimeError: "GeluKernelImpl" not implemented for 'BFloat16'
```
It says it has no such kernel implemented for this dtype.

stas00 commented:
We definitely don't have to do anything about it for now. We will cross that bridge when we come to it.

Until then probably comment out the whole bf16 test, since we might want it down the road when torch implements bf16 support for those kernels. And it's good to have it there for other bf16 tests as an example.

jaketae commented:
I see, thanks for the clarification. I see that it's just a BF16 issue. I'll comment out the test for now. Thanks!

stas00 reviewed (approved):
This is good to go now

stas00 commented on megatron/arguments.py:
```diff
@@ -313,6 +313,10 @@ def _add_network_size_args(parser):
                        help='Define position embedding type ("absolute" | "rotary"). "absolute" by default.'
                        )
+    group.add_argument('--glu-activation', type=str,
+                       choices=["liglu", "geglu", "reglu", "swiglu"],
```
just another possible suggestion to avoid future discrepancies:
```
choices=GLU_ACTIVATIONS.keys()
```
after importing it.

stas00 replied:
but it's fine to leave it as is too for now. doesn't have to be perfect ;)

jaketae replied:
It's a pretty quick fix, so why not? Just committed! Also, thanks for fixing the typo.

stas00 replied:
Hmm, it broke things:
```
___________________________________________ ERROR collecting Megatron-DeepSpeed-master/tests/test_preprocessing.py _____________________________________________
ImportError while importing test module '/mnt/nvme1/code/huggingface/Megatron-DeepSpeed-master/tests/test_preprocessing.py'.
Hint: make sure your test modules/packages have valid Python names.                                                                                              
Traceback:                                                                                                                                                       
/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/importlib/__init__.py:127: in import_module                                                                    
    return _bootstrap._gcd_import(name[level:], package, level)                                                                                                  
tests/test_preprocessing.py:21: in <module>                                                                                                                      
    from megatron.testing_utils import (                                                                                                                         
megatron/__init__.py:28: in <module>                                                                                                                             
    from .global_vars import get_args                                                                                                                            
megatron/global_vars.py:25: in <module>
    from .arguments import parse_args
megatron/arguments.py:25: in <module>
    from megatron.model.glu_activations import GLU_ACTIVATIONS
megatron/model/__init__.py:18: in <module>
    from .distributed import DistributedDataParallel
megatron/model/distributed.py:22: in <module>
    from megatron import get_args
E   ImportError: cannot import name 'get_args' from partially initialized module 'megatron' (most likely due to a circular import) (/mnt/nvme1/code/huggingface/Megatron-DeepSpeed-master/megatron/__init__.py)
```

stas00 replied:
ok, I have a fix, will make a separate PR

stas00 replied:
fixed in https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/72

stas00 replied:
hoping we will have a CI soon - will work on it next week.

jaketae commented:
@stas00 Shouldn't have merged this too quickly, thanks for the quick fix!

stas00 commented:
All is good. It was my unvalidated suggestion that broke it. 

Once we get a CI, it'll be much easier.
